### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by jasontaylordev**
Added support for simple and role-based authorization.

Simple authorization is supported by applying the `[Authorize]` attribute to a command or query. For example:

```csharp
[Authorize]
public class GetTodosQuery : IRequest<TodosVm> { }
```

The above query is accessible to any authenticated user.

Role-based authorization is supported by specifying the relevant role(s). For example, the following code limits the query access to members of the Admin role:

```csharp
[Authorize(Roles = "Admin"]
public class GetTodosQuery : IRequest<TodosVm> { }
```

You can also specify multiple roles as a comma-separated list:

```csharp
[Authorize(Roles = "Admin, SuperUser"]
public class GetTodosQuery : IRequest<TodosVm> { }
```

The above query would only be accessible by users who are a member of the Admin role or the SuperUser role.

If you apply multiple attributes then a user must be a member of all the roles specified:

```csharp
[Authorize(Roles = "Admin")]
[Authorize(Roles = "SuperUser")]
public class GetTodosQuery : IRequest<TodosVm> { }
```

Only accessible to users who are a member of both the Admin role and the SuperUser role.

The above implementation is based on the ASP.NET Core implementation; https://docs.microsoft.com/en-us/aspnet/core/security/authorization/roles?view=aspnetcore-3.1#adding-role-checks. This should make it easier for users already familiar with the ASP.NET Core approach.

### CODE REVIEW

**Comment by jasontaylordev** [comment]
> My first ever github comment on code! I'm sorry if I'm wrong but I think this should be testing for:
> 
> ```cs
> if (authorizeAttributes.Any()) 
> ```
> 
> instead of
> 
> ```cs
> if(authorizeAttributes.Any() == false)
> ```

Thanks, fixed. 👍

**Comment by MCCshreyas** [comment]
Waiting for to get it merged! 😇

**Comment by GFoley83** [comment]
Any thoughts on the comments here Jason: https://github.com/jasontaylordev/CleanArchitecture/issues/56#issuecomment-625549415

Basically, adding some additional logic to the `AuthorizationBehaviour` and `AuthorizeAttribute` to check for claims or policies. It would make the attribute a hell of a lot more usable for very little code. Plus roles are being slowly phased out in favor of role claims.

```cs
[Authorize(Roles = "Admin")]
[Authorize(ClaimType = "MarketingTeamMember", ClaimValue = "true")]
[Authorize(Policy= "SomeAlreadyDefinedPolicy")]
```

Claims check logic:

```cs
public async Task<Result> UserHasClaimAsync(string userId, string type, string value)
{
    var user = _userManager.Users.SingleOrDefault(u => u.Id == userId);

    if (user == null || !(await _userManager.GetClaimsAsync(user)).Any(i => i.Type == type && i.Value == value))
    {
        return Result.Failure(new[] { $"User: {userId} does not have claim: \"{type}\"." });
    }

    return Result.Success();
}

Policy check logic:

```cs
// inject: IAuthorizationService authorizationService
// https://stackoverflow.com/a/43231943/654708
var isAuthorized = await authorizationService.AuthorizeAsync(User, "SomeAlreadyDefinedPolicy");
```

File: src/Application/DependencyInjection.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ public static IServiceCollection AddApplication(this IServiceCollection services
             services.AddAutoMapper(Assembly.GetExecutingAssembly());
             services.AddValidatorsFromAssembly(Assembly.GetExecutingAssembly());
             services.AddMediatR(Assembly.GetExecutingAssembly());
+            services.AddTransient(typeof(IPipelineBehavior<,>), typeof(AuthorizationBehaviour<,>));
             services.AddTransient(typeof(IPipelineBehavior<,>), typeof(PerformanceBehaviour<,>));
             services.AddTransient(typeof(IPipelineBehavior<,>), typeof(ValidationBehaviour<,>));
             services.AddTransient(typeof(IPipelineBehavior<,>), typeof(UnhandledExceptionBehaviour<,>));
```
[the remark this answers is not in the document]
Comment (jasontaylordev) [reply src/Application/DependencyInjection.cs]: Thanks this is fixed

**Comment by jasontaylordev** [comment]
Hey @GFoley83 - I've added support for policy-based authorisation, which could also support claims through policies. Check `PurgeTodoListsCommand` as an example.

**Comment by litan1106** [comment]
@jasontaylordev how does the authorization effects the worker?

does that mean we need to run all commands with [Authorize attribute with user?

```[Authorize(Roles = "Admin, SuperUser"]
public class GetTodosQuery : IRequest<TodosVm> { }```

```await RunAsDefaultUserAsync();
var command = new CreateTodoListsCommand();
```

**Pull request merged**